src/plugins/FetchStickerDataPlugin.ts

```diff
@@ -1,150 +1,124 @@
+    head: '>'
+  });
+
+  await requestQueue.addAll(stickerPackEntries.map(([id, meta], index) => {
+    return async () => {
+      const cachePath = path.resolve(cacheDir, `${id}.json`);
+      const cacheHasPack = await fs.pathExists(cachePath);
+      let stickerPackPartial: StickerPackPartial;
+
+      if (cacheHasPack) {
+        stickerPackPartial = await fs.readJson(cachePath);
+      } else {
+        const stickerPackManifest = await getStickerPackManifest(id, meta.key);
+
+        stickerPackPartial = {
+          meta: {...meta, id},
+          // To keep the size of the generated JSON file small, only extract
+          // the properties from the manifest that we need to generate a list
+          // of search results.
+          manifest: R.pick(['title', 'author', 'cover'], stickerPackManifest)
+        };
+
+        await fs.writeJson(cachePath, stickerPackPartial);
+      }
+
+      stickerPackPartials.push(stickerPackPartial);
+      bar.tick();
+    };
+  }));
+
+  console.log('[FetchStickerDataPlugin] Done.\n');
+
+  return stickerPackPartials;
 }
 
+
 /**
- * Query the manifest for all known sticker packs.
+ * Options accepted by FetchStickerDataPlugin.
  */
-async function getAllStickerPacks(): Promise<StickerPackManifestsJson> {
-  return Promise.all(R.map(async ([id, value]) => {
-    try {
-      // Signal's service will ignore our request if we fetch
-      // too many packs too rapidly so we stagger our timing.
-      await randomDelay(0, 1000);
-      const manifest = await getStickerPack(id, value.key);
-      return {
-        meta: {id, ...value},
-        manifest: {
-          title: manifest.title,
-          author: manifest.author,
-          cover: manifest.cover
-        }
-      };
-    } catch (err) {
-      throw new ErrorWithCode(err.code, `[getAllStickerPacks] ${err.message}`);
-    }
-  }, Object.entries(stickers as StickerPackJson)));
+export interface FetchStickerDataPluginOptions {
+  inputFile: string;
+  outputFile: string;
 }
 
+
 export default class FetchStickerDataPlugin {
-  constructor({filename}) {
-    this.filename = filename;
+  /**
+   * Name of the input file, which should be a YAML document consisting of an
+   * array of StickerPackMetadata objects.
+   */
+  inputFile: string;
+
+  /**
+   * Name of the output file, which will be a JSON document containing an array
+   * of StickerPackPartial objects.
+   */
+  outputFile: string;
+
+
+  constructor({inputFile, outputFile}: FetchStickerDataPluginOptions) {
+    this.inputFile = inputFile;
+    this.outputFile = outputFile;
   }
 
-  apply(compiler) {
-    compiler.plugin('emit', async (compilation, done) => {
-      const json = JSON.stringify(await getAllStickerPacks());
-      compilation.assets[this.filename] = {
+
+  apply(compiler: webpack.Compiler) {
+    compiler.hooks.emit.tapPromise('FetchStickerDataPlugin', async compilation => {
+      const json = JSON.stringify(await getAllStickerPacks(this.inputFile), null, 2);
```

Comment:
Note, I had originally done the `null, 2` thing to pretty format the JSON, but the resulting file is ~35% smaller without this extra whitespace. I think that's a pretty big win for something that's not intended to be human readable/edited anyhow. In reality I'm not sure how much difference it makes after compression and such are applied when it's being served up.

Reply:
Yeah, it looks like even after gzip, we save 5-10% without whitespace, which is worth not having it I think.

Fixed in 2474539.